File: packages/cli/src/drift-diff.ts
export function diffByName<T>(
  expectedItems: readonly T[],
  actualItems: readonly T[],
  getName: (item: T) => string,
  getShape: (item: T) => string
): { missing: string[]; extra: string[]; changed: string[] } {
  const expected = new Map(expectedItems.map((item) => [getName(item), getShape(item)]))
  const actual = new Map(actualItems.map((item) => [getName(item), getShape(item)]))
  const missing: string[] = []
  const extra: string[] = []
  const changed: string[] = []

  for (const [name, expectedShape] of expected.entries()) {
    const actualShape = actual.get(name)
    if (!actualShape) {
      missing.push(name)
      continue
    }
    if (actualShape !== expectedShape) {
      changed.push(name)
    }
  }

  for (const name of actual.keys()) {
    if (!expected.has(name)) {
      extra.push(name)
    }
  }

  return { missing, extra, changed }
}

export function diffSettings(
  expectedSettings: Record<string, string | number | boolean>,
  actualSettings: Record<string, string>
): string[] {
  const settingKeys = Object.keys(expectedSettings).sort()
  const diffs: string[] = []
  for (const key of settingKeys) {
    const left = key in expectedSettings ? String(expectedSettings[key]) : ''
    const right = key in actualSettings ? String(actualSettings[key]) : ''
    if (left !== right) {
      diffs.push(key)
    }
  }
  return diffs
}

export function diffNamedShapeMaps(
  expected: ReadonlyMap<string, string>,
  actual: ReadonlyMap<string, string>
): string[] {
  const keys = [...new Set([...expected.keys(), ...actual.keys()])].sort()
  const diffs: string[] = []
  for (const key of keys) {
    if ((expected.get(key) ?? '') !== (actual.get(key) ?? '')) {
      diffs.push(key)
    }
  }
  return diffs
}

File: chkit_python/src/chkit/cli/commands/drift_diff.py
"""SQL fragment / named-shape diffing helpers used by drift compare.

1:1 port of ``packages/cli/src/commands/drift/diff.ts``.
"""

from __future__ import annotations

from collections.abc import Callable, Mapping, Sequence
from dataclasses import dataclass
from typing import TypeVar

_T = TypeVar("_T")


@dataclass(frozen=True, slots=True)
class DiffByNameResult:
    missing: list[str]
    extra: list[str]
    changed: list[str]


def diff_by_name(
    expected_items: Sequence[_T],
    actual_items: Sequence[_T],
    get_name: Callable[[_T], str],
    get_shape: Callable[[_T], str],
) -> DiffByNameResult:
    """Bucket items into missing / extra / changed by name + shape fingerprint."""
    expected = {get_name(item): get_shape(item) for item in expected_items}
    actual = {get_name(item): get_shape(item) for item in actual_items}
    missing: list[str] = []
    extra: list[str] = []
    changed: list[str] = []

    for name, expected_shape in expected.items():
        actual_shape = actual.get(name)
        if actual_shape is None:
            missing.append(name)
            continue
        if actual_shape != expected_shape:
            changed.append(name)

    extra = [name for name in actual if name not in expected]

    return DiffByNameResult(missing=missing, extra=extra, changed=changed)


def diff_settings(
    expected_settings: Mapping[str, str | int | float | bool],
    actual_settings: Mapping[str, str],
) -> list[str]:
    """Return keys whose string-cast value differs between expected and actual."""
    diffs: list[str] = []
    for key in sorted(expected_settings):
        left = str(expected_settings[key])
        right = str(actual_settings.get(key, ""))
        if left != right:
            diffs.append(key)
    return diffs


def diff_named_shape_maps(
    expected: Mapping[str, str],
    actual: Mapping[str, str],
) -> list[str]:
    """Return keys whose normalized shape strings differ between the two maps."""
    keys = sorted(set(expected) | set(actual))
    return [key for key in keys if expected.get(key, "") != actual.get(key, "")]
